upstairs/src/block_req.rs

```diff
@@ -2,69 +2,40 @@
 use super::*;
 use tokio::sync::oneshot;
 
-/**
- * Couple a BlockOp with a notifier for calling code. This uses a single-use
- * channel to send the result of a particular operation, and is meant to be
- * paired with a BlockReqWaiter.
- */
-#[must_use]
+/// Wrapper to contain a `BlockOp`
+///
+/// The `BlockReq` used to have more fields, but now it's down to just this one;
+/// we can remove it at a later point.
 #[derive(Debug)]
 pub(crate) struct BlockReq {
     pub op: BlockOp,
-    pub res: BlockRes,
-}
-
-#[must_use]
-#[derive(Debug, PartialEq)]
-pub(crate) struct BlockReqReply {
-    /// We return the buffer (if passed in) so the host can reuse it
-    pub buffer: Option<Buffer>,
-
-    /// Actual result of the Crucible operation
-    pub result: Result<(), CrucibleError>,
 }
 
 #[must_use]
 #[derive(Debug)]
-pub(crate) struct BlockRes(Option<oneshot::Sender<BlockReqReply>>);
+pub(crate) struct BlockRes<T = (), E = CrucibleError>(
+    Option<oneshot::Sender<Result<T, E>>>,
+);
 
-impl BlockRes {
+impl<T, E> BlockRes<T, E> {
     /// Consume this BlockRes and send Ok to the receiver
-    pub fn send_ok(self) {
-        self.send_result(None, Ok(()))
-    }
-
-    /// Consume this BlockRes and send Ok (with data) to the receiver
-    pub fn send_ok_with_buffer(self, buffer: Buffer) {
-        self.send_result(Some(buffer), Ok(()))
+    pub fn send_ok(self, t: T) {
+        self.send_result(Ok(t))
     }
 
     /// Consume this BlockRes and send an Err to the receiver
-    pub fn send_err(self, e: CrucibleError) {
-        self.send_result(None, Err(e))
-    }
-
-    /// Consume this BlockRes and send Err (with data) to the receiver
-    pub fn send_err_with_buffer(self, buffer: Buffer, e: CrucibleError) {
-        self.send_result(Some(buffer), Err(e))
+    pub fn send_err(self, e: E) {
+        self.send_result(Err(e))
     }
 
     /// Consume this BlockRes and send a Result to the receiver
-    fn send_result(
-        mut self,
-        buffer: Option<Buffer>,
-        result: Result<(), CrucibleError>,
-    ) {
+    pub(crate) fn send_result(mut self, result: Result<T, E>) {
         // XXX this eats the result!
```

Comment:
Should this XXX comment remain?

Reply:
It hasn't changed from before, so I decided to leave it.